### ISSUE TITLE
Support deprecating db methods and removing after a major revision

### ISSUE THREAD

**Issue by djmitche** [taskcluster/taskcluster#3328]
We recently introduced a little twist to the compatibility matrix: service upgrades must jump by at most one major version at a time.  So going from v50.1.2 to v52.0.5 is not guaranteed not to break.

That means that if a method is deprecated in v50.1.2, then the method needs to remain functional throughout the v51.x.y series, as a deployment might upgrade from v50.1.2 to v51.x.y an expect things to work.  But that function can be removed completely in v52.

So basically, such functions need to wait for two major versions before they can be removed.  Let's find a way to ensure that, either with a test or in a way that's obvious in reviews.  That might be as simple as changing

```yaml
deprecated: true
```
to
```yaml
deprecated: until v52
```
where v52 is two major versions ahead of the current version when this change lands.

**Comment by helfi92** [comment]
I understand how going from v50.1.2 to v52.0.5 may break I'm not sure how this entails that going from v51.x.y to v52 won't break. Any clarifications would be good. Thanks!

**Comment by djmitche** [comment]
Imagine:

 * old versions of TC have a `get_widgets` function
 * TC v50.1.2 introduces a new `inventory` table and a `get_inventory` method, with `get_widgets` rewritten to call `get_inventory(product_in := "widget", ..)` as a temporary workaround.  All JS in this version is updated to use `get_inventory` and not `get_widgets`.
 * TC v51.x.y doesn't modify this part of the DB, so `get_widgets` and `get_inventory` both still work.
 * TC v52.0.0 further modifies the inventory support, in the process dropping support for the deprecated `get_widgets` method.

So TC versions up through v50.1.1 call `get_widgets` and everything after calls `get_inventory`.

Our compatibility guarantee says
> Taskcluster maintains the invariant that older code is compatible with newer databases, so an existing services deployment will continue to function after a database upgrade is committed. This compatibility is limited to major Taskcluster versions: code in version vN.x.y will function against database versions released in v(N+1).x.y.

So in this case version v50.1.1 must function against all v51.x.y, and it does because all of those versions still implement a working `get_widgets`.

The reason that upgrading the DB from v51.x.y to v52 will work is that all v51.x.y use `get_inventory` and not `get_widgets`, so dropping support for `get_widgets` in the DB has no effect.

This scheme is actually a little conservative, and there are cases where we could get away with only waiting one major version, but it's simpler to just say "two major versions" every time.

**Comment by djmitche** [comment]
The thing I implemented addresses that "conservative" note -- a deprecation that lands in vX.0.0 can be broken in v(X+1).0.0.  See the comments in the PR.

**Comment by helfi92** [comment]
Thanks for the clarification.
